Some minor type stub improvements

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118870
* __->__ #118529

I was just playing around with improving the typing of symbolic_shapes. The PR is not "complete" but I in particular wanted to get feedback on whether or not people liked making ValueRanges Generic; it seems that distinguishing if you have an Expr ValueRange or a SympyBoolean ValueRange is a lot of trouble for downstream. Using TypeGuard, we can perform refinements on the generic parameter inside methods, although we still have to cast back to ValueRange[T] due to https://github.com/python/mypy/issues/14425#issuecomment-1914852707

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

ezyang committed:
Some minor type stub improvements

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -96,40 +109,42 @@ def tighten(self, other) -> "ValueRanges":
-        if self.is_bool:
-            range = ValueRanges(sympy.Or(self.lower, other.lower), sympy.And(self.upper, other.upper))
+        if vr_is_bool(self):
+            return cast(ValueRanges[T], ValueRanges(sympy.Or(self.lower, other.lower), sympy.And(self.upper, other.upper)))
```
While I like the idea in theory, this `cast` is a bit awkward. I guess it's fine if we just do it in two places really...

ezyang replied:
Whenever you do a refinement using `vr_is_bool` and then refine the return type based on this refinement, we have to do it. I only did it in two places because that's only as far as I got.

lezcano replied:
Right. It looks a bit clunky, but I guess that may come from me not being used to typing in Python. I'll let the other ones weigh in. I'm probably +0 on this.

Skylion007 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -16,6 +19,8 @@
 
 __all__ = ["ValueRanges", "ValueRangeAnalysis", "bound_sympy"]
 
+T = TypeVar('T', sympy.Expr, SympyBoolean)
```
Shouldn't this be a private TypeVar? (leading underscore)

ezyang replied:
Oh, I don't actually know! When should a TypeVar ever be public lol

Skylion007 replied:
When someone might one use it downstream code or their own type annotations (ie. extending an abstract base class or something).

ezyang replied:
That's so strange to me though, I always expected you would just redefine a TypeVar if you needed to do that haha

ezyang replied:
Ughhh adding the leading underscore is so uglyyyyy

Skylion007 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -96,40 +109,42 @@ def tighten(self, other) -> "ValueRanges":
         else:
-            range = ValueRanges(sympy.Min(self.lower, other.lower), sympy.Max(self.upper, other.upper))
-        return range
+            raise AssertionError("impossible")
```
```suggestion
            typing_extensions.assert_never(self)
```
Is how you are suppose to handle unreachable code with typecheckers.

ezyang replied:
Can't do this, mypy won't be able to prove that the code is unreachable because it doesn't know that vr_is_bool and vr_is_expr covers all cases

Skylion007 commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -96,40 +109,42 @@ def tighten(self, other) -> "ValueRanges":
         else:
-            range = ValueRanges(sympy.Max(self.lower, other.lower), sympy.Min(self.upper, other.upper))
-        return range
+            raise AssertionError("impossible")
```
```suggestion
            typing_extensions.assert_never(self)
```

ezyang committed:
Update on "Some minor type stub improvements"


I was just playing around with improving the typing of symbolic_shapes. The PR is not "complete" but I in particular wanted to get feedback on whether or not people liked making ValueRanges Generic; it seems that distinguishing if you have an Expr ValueRange or a SympyBoolean ValueRange is a lot of trouble for downstream. Using TypeGuard, we can perform refinements on the generic parameter inside methods, although we still have to cast back to ValueRange[T] due to https://github.com/python/mypy/issues/14425#issuecomment-1914852707

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Some minor type stub improvements"


I was just playing around with improving the typing of symbolic_shapes. The PR is not "complete" but I in particular wanted to get feedback on whether or not people liked making ValueRanges Generic; it seems that distinguishing if you have an Expr ValueRange or a SympyBoolean ValueRange is a lot of trouble for downstream. Using TypeGuard, we can perform refinements on the generic parameter inside methods, although we still have to cast back to ValueRange[T] due to https://github.com/python/mypy/issues/14425#issuecomment-1914852707

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
